### ISSUE TITLE
Update code to be TRAPI 1.0 compliant in NewFmt branch

### ISSUE THREAD

**Issue by finnagin** [RTXteam/RTX#1184]
Creating an issue to track the TRAPI 1.0 update since we do not appear to have one yet.

**Comment by dkoslicki** [comment]
Helpful doc [here](https://docs.google.com/document/d/1-IlEFcc7eVJx6pL2_jHC71kQJS4Bo5rSlrUj5lNPLms/edit?usp=sharing)

**Comment by chunyuma** [comment]
Question: does `kg.nodes` still have an attribute named `qnode_ids` (probably it should be named `qnode_keys`)? In other words, if I want to know what is the `qnode_keys` for a node in `kg.nodes`, can I still access it from `kg.nodes[key].qnode_keys`?

**Comment by edeutsch** [comment]
This was a custom feature that we added outside the schema. So I think this is up to @amykglen on the best way to handle this?

**Comment by edeutsch** [comment]
So I believe we decided that all instances of node_id become node_key and edge_id becomes edge_key, is this agreed:
```
overlay(action=compute_ngd, source_qnode_id=n00, target_qnode_id=n01, virtual_relation_label=N1)
```
becomes
```
overlay(action=compute_ngd, source_qnode_key=n00, target_qnode_key=n01, virtual_relation_label=N1)
```
in the DSL? right? or?

i.e. re start referring to all n00 and e00 types things as "keys" everywhere?

**Comment by edeutsch** [comment]
Expand, too:
```
      - expand(edge_key=e00, kp=BTE, continue_if_no_results=true)
      - overlay(action=overlay_clinical_info, observed_expected_ratio=true, source_qnode_key=n00, target_qnode_key=n01, virtual_relation_label=CP1)
```
?

**Comment by chunyuma** [comment]
@edeutsch, I agree that we should use `key` to replace the original `id`. I would also suggest to change all `source` to `subject` (eg. `source_qnode_key` to `subject_qnode_key`) and all `target` to `object` (eg. `target_qnode_key` to `object_qnode_key`).

**Comment by edeutsch** [comment]
ah, good point. I would agree. others?

**Comment by amykglen** [comment]
that all seems to make sense to me (including changing `node.qnode_ids` to `node.qnode_keys`).

**Comment by chunyuma** [comment]
Hi @edeutsch, to get `message`, expect for using `message = response.envelope.message`, is it possible to use the following code:

```
query = {"previous_message_processing_plan": {"processing_actions": [
    "create_message",
    f"add_qnode(id={query_node_curie}, category={source_type}, key=FET_n00)",
    f"add_qnode(category={adjacent_type}, key=FET_n01)",
    f"add_qedge(subject=FET_n00, object=FET_n01, key=FET_e00)",
    f"expand(edge_key=FET_e00,kp={kp})",
    "return(message=true, store=false)"
]}}

araxq = ARAXQuery()
result = araxq.query(query)
message = araxq.message
```

**Comment by edeutsch** [comment]
Yes, I think that is right. But ARAXQuery isn't quite finished in the conversion yet, so this probably doesn't quite work yet? I will work on that next.

**Comment by finnagin** [comment]
@edeutsch it looks like the OpenApi Edge class has far fewer attributes than the Swagger one.
i.e.
```
#swagger class:
self.attribute_map = {
            'id': 'id',
            'type': 'type',
            'relation': 'relation',
            'source_id': 'source_id',
            'target_id': 'target_id',
            'is_defined_by': 'is_defined_by',
            'defined_datetime': 'defined_datetime',
            'provided_by': 'provided_by',
            'confidence': 'confidence',
            'weight': 'weight',
            'publications': 'publications',
            'evidence_type': 'evidence_type',
            'qualifiers': 'qualifiers',
            'negated': 'negated',
            'edge_attributes': 'edge_attributes',
            'qedge_ids': 'qedge_ids'
        }
```
is now
```
#OpenApi class
self.attribute_map = {
            'predicate': 'predicate',
            'relation': 'relation',
            'subject': 'subject',
            'object': 'object',
            'attributes': 'attributes'
        }
```

Do you have any guidance on how I can go about converting this old Edge class to the new one when it is initiated in the code like so?
```
edge = Edge(id=id, type=edge_type, relation=relation, source_id=source_id, target_id=target_id,
                            is_defined_by=is_defined_by, defined_datetime=defined_datetime, provided_by=provided_by,
                            confidence=confidence, weight=weight, edge_attributes=[edge_attribute], qedge_ids=qedge_ids)
```

**Comment by edeutsch** [comment]
yes, thanks. So all those other things will need to go into the attributes[ ] slot. For each of those, create an Attribute object and then append it to the list of attributes. more tedious for sure, but we decided that there was not a good way to agree which of those should be first-class attributes and which should be a second-class attribute, and so they're all second-class attributes.

Make sense? Questions?

**Comment by finnagin** [comment]
That makes sense. Thank you!

**Comment by finnagin** [comment]
@edeutsch  How do we add an option_group_id to the new QEdge class in OpenAPi?

**Comment by edeutsch** [comment]
I'm not certain I really understand the question. It would probably be tidiest to use the messenger.add_qedge(response) method call. Is that what you mean? Alternatively, if you already have a qnode object and you just want to add it, you could do:

qnode.option_group_id = '1'

I suspect I'm not understanding the question.

**Comment by finnagin** [comment]
In the swagger api the QEdge class used to have negated, option_group_id, ect. However, the new OpenApi version does not. 
i.e.
```
# from swagger api class
self.attribute_map = {
            'id': 'id',
            'type': 'type',
            'relation': 'relation',
            'source_id': 'source_id',
            'target_id': 'target_id',
            'negated': 'negated',
            'option_group_id': 'option_group_id',
            'exclude': 'exclude'
        }
```
is now
```
# from the OpenApi class
self.attribute_map = {
            'predicate': 'predicate',
            'relation': 'relation',
            'subject': 'subject',
            'object': 'object'
        }
```
But unlike with the Edge class there is no Attribute list in order to encode additional attributes. So I was wondering what the best way to go about including those attributes is. If adding them manually like you said is the best way I can do that.

**Comment by finnagin** [comment]
I guess I'm wondering should we expand the OpenApi model to be more lake swagger here?

**Comment by edeutsch** [comment]
ah, I understand now. I suppose these were custom attributes that we added in the previous data model. and it looks like I neglected to propagate them to the latest revision.
So it looks like I need to propagate 'option_group_id' and 'exclude' to the latest data model?
Did I forget anything else?

**Comment by amykglen** [comment]
`Node.qnode_keys` and `Edge.qedge_keys`?

**Comment by edeutsch** [comment]
aha, right. will add.
Anything else?

**Comment by edeutsch** [comment]
ooo. wait. That one if potentially a problem.
Node and Edge are officially
      additionalProperties: false
in the TRAPI 1.0 schema. (QNode and QEdge are okay)
So I cannot legally add them.
Can you just quietly stuff them in without telling anyone?
Then if they're not in the schema they won't get written out in the output and no one will know what we put them in there temporarily.
Can you try that?

**Comment by amykglen** [comment]
ah, ok. yeah, I believe that works. slightly inconvenient in that you can't specify them when initializing a Node/Edge (e.g., trying to do `Node(name="Parkinsons", type="disease", qnode_keys=["n00"])` throws an error. but instead you can just set the qnode_keys after initializing.

**Comment by edeutsch** [comment]
I think we'll need to live with that because we cannot extend Node and Edge. All information we want to add MUST go into attributes[ ]. Unless we sneak it in separately, and then it won't be written back out by the framework.

**Comment by edeutsch** [comment]
okay, the exclude and option_group_if fields are part of the new models in NewFmt now. Let me know if you still have trouble with it.

**Comment by edeutsch** [comment]
We've made a lot of progress, but the following test:
```
python3 ARAQ_query.py 0
```
fails on the ranker and I have not been able to understand the code enough to fix. Need help.

**Comment by edeutsch** [comment]
I removed the ranker for further testing, and it looks like filter_results isn't working either. Even when told to filter results at 5, all 380+ results come back.

**Comment by finnagin** [comment]
@edeutsch filter results is working now. It passes test_filter_results.py when the ranker is disabled.

I'm going to take a pass at trying to fix the ranker.

**Comment by finnagin** [comment]
@edeutsch I believe ranker is working now. I fixed the error that was crashing it and haven't had any other issues yet.

**Comment by edeutsch** [comment]
@finnagin great, thanks! The system is a little closer to working!

**Comment by chunyuma** [comment]
The errors for fisher_exact_test and predict_drug_treats_disease are fixed. All tests for these two functions are passed.

**Comment by edeutsch** [comment]
excellent, many thanks! all non runslow tests are now passing!

**Comment by finnagin** [comment]
NewFmt has now been merged into master. I think that means we are now officially TRAPI 1.0 compliant!

**Comment by amykglen** [comment]
when `overlay` adds edges to the KG, should it be using a particular prefix in the predicates, even if the predicate doesn't exist in biolink? i.e., are we supposed to invent our own prefix, like `ARAX:has_normalized_google_distance_with`?

![Screen Shot 2021-01-25 at 5 59 27 PM](https://user-images.githubusercontent.com/49423686/105789418-40cdc800-5f37-11eb-8e92-e3460361ba9e.png)

**Comment by edeutsch** [comment]
In fact, the schema *requires* that the biolink: prefix be used: https://github.com/NCATSTranslator/ReasonerAPI/blob/c77d48eb9e8208ad400d890bf706f60db48d4553/TranslatorReasonerAPI.yaml#L574

I believe the correct way to do things it propose the term to BioLink, although we're sort of free to start using it anyway. This seems a little shady, but I believe I heard Chris Mungall give his approval to this approach.

**Comment by finnagin** [comment]
Ok, I'll add the biolink prefix to the overlay edges then. So, `has_normalized_google_distance_with` will become `biolink:has_normalized_google_distance_with`. Sound good?

**Comment by finnagin** [comment]
Should I also make it Pascal case so it matches the other edges in our graph? i.e. `has_normalized_google_distance_with` will become `biolink:HasNormalizedGoogleDistanceWith`.

**Comment by edeutsch** [comment]
Node categories are PascalCase, but edge predicates are snake_case so biolink:has_normalized_google_distance_with is correct.

**Comment by finnagin** [comment]
Okay I'll do that then, thanks!

**Comment by amykglen** [comment]
can this issue be closed? I'd say we're TRAPI 1.0 compliant?

**Comment by edeutsch** [comment]
Yes, let's close it! Just in time to open an issue for implementing TRAPI 1.1, wheee!
